File: T2/bounds.py
#!/usr/bin/env python3

from math import cos, pi

def cauchy(func, tol):
    def cauchy_max(func, tol):
        n = len(func)
        max_rad = (abs(func[-1] / func[0])) ** (1/n-1)
        while True:
            init_rad = max_rad
            max_rad = sum(abs(func[j] / func[0]) * (init_rad ** (n - j)) for j in range(1, n)) ** (1/n)
            if (abs(max_rad - init_rad) < tol):
                return max_rad

    return (cauchy_max(func, tol), cauchy_max(func[::-1], tol))
# def cauchy_max(func, tol):
#     n = len(func)
#     max_rad = (abs(func[-1] / func[0])) ** (1/n-1)
#     while True:
#         init_rad = max_rad
#         max_rad = sum(abs(func[j] / func[1]) * (init_rad ** (n - j)) for j in range(1, n)) ** (1/n)
#         if (abs(max_rad - init_rad) < tol):
#             break

#     return max_rad

# def cauchy_min(func, tol):
#     n = len(func)
#     min_rad = abs(func[0] / func[-1]) ** (1/n-1)
#     while True:
#         init_rad = min_rad
#         min_rad = 1 / sum(abs(func[j] / func[-1]) * init_rad ** (j) for j in range(1, n)) ** (1/n)
#         if (abs(init_rad - min_rad) < tol):
#             break

#     return min_rad

def kojima(func):
    def kojima_max(func):
        queue = [abs(func[i] / func[0]) ** (1 / i) for i in range(1, len(func))]
        return sum(sorted(queue)[-2:])

    return (kojima_max(func), 1 / kojima_max(func[::-1]))

def track(f, tol):
    n = len(f)

    max_rad = [None] * 3
    min_rad = [None] * 3

    max_rad[0] = 1 + max(abs(i) for i in f) / abs(f[0])
    min_rad[0] = 1 / (1 + max(abs(i) for i in f)) / abs(f[-1])

    print(min_rad[0])

    max_rad[1] = cauchy_max(f, tol)
    max_rad[2] = kojima_max(f)
    
    print(max_rad[1])

    min_rad[1] = cauchy_min(f, tol)
    min_rad[2] = kojima_min(f, min_rad[1])

    rmax = min(max_rad)
    rmin = max(min_rad)

    mean = (rmax + rmin) / 2
    
    a = mean * cos(pi/4)
    b = (-1) * a
    xi = complex(a, b)

    return xi

File: T2/octave/bounds.m
function xI = bounds(n,a)
    % Max bound
    max_rad(1) = 1 + max(abs(a(2:n+1))) / abs(a(1));
    min_rad(1) = 1 / (1 + max(abs(a(1:n))) / abs(a(n+1)));

    % Cauchy bound
    % max:
    rI = abs(a(n+1) / a(1)) ^ (1/n);
    for k = 1 : 30
        t_sum = abs(a(n+1) / a(1));
        for i = 2 : n
            t_sum = t_sum + abs(a(i) / a(1)) * rI ^ (n-i+1);
        end
        max_rad(2) = t_sum ^ (1/n);
        rI = max_rad(2);
    end
    %min:
    rI = abs(a(1) / a(n+1)) ^ (1/n);
    for k = 1 : 30
        t_sum = abs(a(1) / a(n+1));
        for i = 2 : n
            t_sum = t_sum + abs(a(i) / a(n+1)) * rI ^ (i-1);
        end
        rI = t_sum ^ (1/n);
    end
    min_rad(2) = 1/rI;

    % Kojima bound
    % max:
    for i = 2 : n+1
        q(i) = abs(a(i) / a(1)) ^ (1 / (i-1));
    end
    order = sort(q);
    max1 = order(n+1);
    max2 = order(n);
    max_rad(3) = max1 + max2;
    % min:
    for i = 2 : n+1
        q(i) = abs(a(n-i+2) / a(n+1)) ^ (1 / (i-1));
    end
    order = sort(q);
    max1 = order(n+1);
    max2 = order(n);
    min_rad(3) = 1 / (max1 + max2);

    rmax = min(max_rad);
    rmin = max(min_rad);

    mean_rad = ((rmax+rmin)/2);
    a = mean_rad * cos(pi/4);
    b = -a;
    xI = complex(a,b);
end
